sshd-core/src/main/java/org/apache/sshd/client/session/ClientSessionImpl.java

```diff
@@ -90,16 +88,9 @@ public ClientSessionImpl(ClientFactoryManager client, IoSession ioSession) throw
         authFuture.setAuthed(false);
 
         signalSessionCreated(ioSession);
-        sendClientIdentification();
 
-        KexExtensionHandler extHandler = getKexExtensionHandler();
-        if ((extHandler == null) || (!extHandler.isKexExtensionsAvailable(this, AvailabilityPhase.PREKEX))) {
-            kexState.set(KexState.INIT);
-            sendKexInit();
-        } else {
-            if (log.isDebugEnabled()) {
-                log.debug("<init>({}) delay KEX-INIT until server-side one received", this);
-            }
+        if (sendImmediateIdentification) {
+            initializeKexPhase();
```

Comment:
I don't think this is OK. This will make it impossible to implement any client-side proxy protocol to connect through SOCKS or HTTP CONNECT proxies if `!sendImmediateIdentification`. If sending the client identification is delayed, the proxy protocol must still be started before expecting an answer from the server.

Reply:
Please see the updated code that initializes the proxy connector regardless of whether immediate send is enabled or not.

Reply:
That would work.

Reply:
See updated code in view of your comments (thanks for the feedback)